core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -131,55 +136,98 @@ public BlockInfo getInfo(long blockId) throws IOException {
+  }
+
+  /**
+   * Gets a stream to read the data of a block. This method is primarily responsible for
+   * determining the data source and type of data source. The latest BlockInfo will be fetched
+   * from the master to ensure the locations are up to date. It takes a map of failed workers and
+   * their most recently failed time and attempts to avoid reading from a recently failed worker.
+   *
+   * @param blockId the id of the block to read
+   * @param options the options associated with the read request
+   * @param failedWorkers the map of workers address to most recent failure time
+   * @return a stream which reads from the beginning of the block
+   */
+  public BlockInStream getInStream(long blockId, InStreamOptions options,
+      Map<WorkerNetAddress, Long> failedWorkers) throws IOException {
     // Get the latest block info from master
     BlockInfo info;
     try (CloseableResource<BlockMasterClient> masterClientResource =
              mContext.acquireBlockMasterClientResource()) {
       info = masterClientResource.get().getBlockInfo(blockId);
     }
     List<BlockLocation> locations = info.getLocations();
-    if (locations.isEmpty() && !options.getStatus().isPersisted()) {
+    List<BlockWorkerInfo> blockWorkerInfo = Collections.EMPTY_LIST;
+    // Initial target workers to read the block given the block locations.
+    Set<WorkerNetAddress> workerPool;
+    if (options.getStatus().isPersisted()) {
+      blockWorkerInfo = getEligibleWorkers();
+      workerPool = blockWorkerInfo.stream().map(BlockWorkerInfo::getNetAddress).collect(toSet());
+    } else {
+      workerPool = locations.stream().map(BlockLocation::getWorkerAddress).collect(toSet());
+    }
+    if (workerPool.isEmpty()) {
       throw new NotFoundException(ExceptionMessage.BLOCK_UNAVAILABLE.getMessage(info.getBlockId()));
     }
-    // Determine the data source and the type of data source
+    // Workers to read the block, after considering failed workers.
+    final Set<WorkerNetAddress> workers = handleFailedWorkers(workerPool, failedWorkers);
     // TODO(calvin): Consider containing these two variables in one object
     BlockInStreamSource dataSourceType = null;
     WorkerNetAddress dataSource = null;
-    if (locations.isEmpty()) { // Data will be read from UFS
-      dataSourceType = BlockInStreamSource.UFS;
-      BlockLocationPolicy policy =
-          Preconditions.checkNotNull(options.getOptions().getUfsReadLocationPolicy(),
-              PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
-      GetWorkerOptions getWorkerOptions = GetWorkerOptions.defaults().setBlockId(info.getBlockId())
-          .setBlockSize(info.getLength()).setBlockWorkerInfos(getEligibleWorkers());
-      dataSource = policy.getWorker(getWorkerOptions);
-    } else { // Data will be read from Alluxio, determine which worker and if it is local
+    locations = locations.stream()
+        .filter(x -> workers.contains(x.getWorkerAddress())).collect(toList());
```

Comment:
Could you use a more descriptive variable name (applies in general to lambdas)?

Reply:
Done.